Process Inheritance creates copy of kobject

jmazanec15 opened the pull request:
This PR is the first step to address #228 . Originally, in process_selective_inheritance, the parent gave the child process a pointer to each of its kobjects it wanted to inherit. This produced a problem of having undesired shared state across different processes. For example, in window manager, changes to a kobject in two child processes affected each other.

This PR addresses this problem by instead of calling kobject_addref in process_selective_inherit, it calls kobject_copy. Kobject_copy creates a new kobject for the child and creates a shallow copy of of the parent's k_object that is passed in. *Note*: I am not sure if a new kobject should be initialized inside of the kobject_copy function, or inside of the process_selective_inherit function.

Along with this, I modified how kobject_close and kobject_addref works. I also added a blocking flag for the kobject struct.

Finally, I added a function for each object type called getref that returns the underlying reference count.

dthain commented:
Hmm, sorry, I think this is on the wrong track, there is a simpler way to go about it:

1 - We still want reference counting for kobjects, since they can be dup'd within a single process.  So `kobject_addref` and `kobject_close` should continue to work as before.  That allows you to have a single kobject with refcount 2 pointing to (for example) a window with refcount 1.

2 - All `kobject_copy` needs to do is copy the internal fields, call addref on the underlying object, and return a kobject with refcount 1.  That allows you to have two kobjects each referring to the same underlying item.

3 - The `getref` operations are subtly dangerous:  you don't want to allow a caller to make decisions based on the underlying reference count of an object, since that can change between the time you check and the time you act.  The only two operations should be to add a reference and to close (potentially destroy) an object.

jmazanec15 commented:
The changes you suggested made sense so I just made them. One question on kobject_copy: I still init a new kobject in the function. I do not know if this is the best way to create a new kobject or if it should be created in the process_selective_inherit function.

dthain commented on kernel/kobject.c:
```diff
@@ -226,23 +227,44 @@ int kobject_lookup( struct kobject *kobject, const char *name, struct kobject **
+	(*kdst)->data 			= ksrc->data;
+	(*kdst)->type 			= ksrc->type;
+	(*kdst)->offset 		= ksrc->offset;
+	(*kdst)->tag 				= ksrc->tag;
```
This string is malloc'd so it probably needs to be strdup'd, yes?

dthain commented on kernel/kobject.c:
```diff
@@ -226,23 +227,44 @@ int kobject_lookup( struct kobject *kobject, const char *name, struct kobject **
+	case KOBJECT_PIPE:
+		pipe_addref(ksrc->data.pipe);
+		break;
+	default:
```
As a style point, I would *not* have a default case here.
The compiler can check whether we used all of the switch cases here.
If a new one gets added (and no default is provided) then the compiler will tell us.

dthain commented on kernel/kobject.c:
```diff
@@ -226,23 +227,44 @@ int kobject_lookup( struct kobject *kobject, const char *name, struct kobject **
 }
 
-int kobject_copy( struct kobject *ksrc, struct kobject *kdst )
+int kobject_copy( struct kobject *ksrc, struct kobject **kdst )
```
Suggest that you just return the new kobject as the return value.
Makes the function prettier to use below.

dthain commented on kernel/kobject.c:
```diff
@@ -294,12 +316,7 @@ int kobject_close(struct kobject *kobject)
-	default:
-		return 0;
-	}
+	kobject->blocking = b;
```
This looks like an incomplete solution to the blocking problem.
For the purpose of this PR, just leave blocking the way it is,
and fix the problem thoroughly in a different PR.

dthain commented on kernel/kobject.h:
```diff
@@ -28,6 +28,7 @@ struct kobject {
 	int refcount;
 	int offset;
 	char *tag;
+	int blocking;
```
Ibid.

dthain commented:
Notice that there is a (very) similar string.c in the kernel directory, as well as in the user directory.  The trick is to add strdup to the kernel side as well.

dthain commented:
And keep in mind that `tag` may be null, indicating no tag at all, so you have to check for that case too.  (Probably the reason for the invalid kfree)

jmazanec15 commented:
Those changes made sense so I fixed them. Doing some debugging, the invalid kfree comes from console_delete. I am looking into it now.

jmazanec15 commented:
So, the invalid kfree occurs because console_delete tried to kfree the console_root. This has two problems: 1.) console_root is a global variable, which cannot be freed and 2.) the root console is used by the main program so the refcount should not be 0.

To fix 1, I added a check to make sure that the console_root is never freed in console_delete.
To fix 2, I added console_addref(&console_root) in main.c.

I am not sure if we want console_root to continue to be a global variable, but this solves the invalid kfree error.

dthain commented:
Looks good!
